cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/AbstractInMemoryProgramRunner.java

```diff
@@ -144,14 +145,33 @@ private void monitorComponents() {
         controller.addListener(new AbstractListener() {
           @Override
           public void completed() {
-            if (liveComponents.decrementAndGet()  == 0) {
-              complete();
+            if (liveComponents.decrementAndGet() == 0) {
+              // If at least one of the components failed with an error, then set the state to ERROR
+              if (errorCause != null) {
+                setError(errorCause);
+              } else {
+                complete();
+              }
+            }
+          }
+
+          @Override
+          public void error(Throwable cause) {
+            // Make a note that a component failed with an error
+            errorCause = cause;
+            // If the live component count goes to zero, then set the state of the controller to ERROR
+            if (liveComponents.decrementAndGet() == 0) {
+              setError(cause);
```

Comment:
We can simply call `error(cause)` here. No need for extra single line `setError` method.